tools/internal/autoupdate/registry.go

```diff
@@ -0,0 +1,348 @@
+package autoupdate
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/Masterminds/semver/v3"
+	"github.com/rancher/image-mirror/internal/config"
+	"io"
+	"math"
+	"net/http"
+	"net/url"
+	"os"
+	"regexp"
+	"sort"
+	"strings"
+	"time"
+)
+
+type Registry struct {
+	Images        []AutoupdateImageRef
+	Latest        bool   `json:",omitempty"`
+	LatestEntry   bool   `json:",omitempty"`
+	VersionFilter string `json:",omitempty"`
+}
+
+var httpClient = &http.Client{Timeout: 10 * time.Second}
+
+// DockerHubResponse matches the structure of the Docker Hub API response
+type DockerHubResponse struct {
+	Next    string `json:"next"`
+	Results []struct {
+		Name string `json:"name"`
+	} `json:"results"`
+}
+
+// QuayResponse matches the structure of the Quay.io API response
+type QuayResponse struct {
+	HasAdditional bool `json:"has_additional"`
+	Tags          []struct {
+		Name string `json:"name"`
+	} `json:"tags"`
+}
+
+type SuseTokenResponse struct {
+	AccessToken string `json:"access_token"`
+}
+
+type GenericTagsResponse struct {
+	Tags []string `json:"tags"`
+}
+
+func (r *Registry) GetUpdateImages() ([]*config.Image, error) {
+	githubToken := os.Getenv("GITHUB_TOKEN")
+	allTags, err := r.getImageTags(githubToken, "", 1)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get image tags: %w", err)
+	}
+	versionFilter := regexp.MustCompile(r.VersionFilter)
+
+	var filteredTags []string
+	for _, tag := range allTags {
+		if !versionFilter.MatchString(tag) {
+			continue
+		}
+		filteredTags = append(filteredTags, tag)
+	}
+
+	if len(filteredTags) == 0 {
+		return nil, errors.New("no tags found matching version filter")
+	}
+
+	if r.LatestEntry {
```

Comment:
It might be a good idea to merge the `if` statements for `r.LatestEntry` and `r.Latest` into an `else-if` statement, since they are mutually exclusive. Feel free to disregard.